solana-bridges/solana-client/SolanaClient.sol

```diff
@@ -712,6 +769,250 @@ struct Slot {
+            (vote.timestamp, cursor) = parseUint64LE(buffer, cursor);
+        }
+
+        return (vote, cursor);
+    }
+
+    function parseCompactWord16(bytes memory bs, uint cursor) public pure returns (uint16, uint) {
+        uint8 b0 = uint8(bs[cursor]); cursor++;
+        uint16 w = b0 & 0x7f;
+        if (b0 < (1 << 7))
+            return (w, cursor);
+
+        uint8 b1 = uint8(bs[cursor]); cursor++;
+        w |= uint16(b1 & 0x7f) << 7;
+        if (b1 < (1 << 7))
+            return (w, cursor);
+
+        uint8 b2 = uint8(bs[cursor]); cursor++;
+        w |= uint16(b2 & 0x03) << 14;
+        if (b2 < (1 << 2))
+            return (w, cursor);
+
+        revert("Invalid Compact-u16");
+    }
+
+    uint256 constant voteTag = 2;
+    uint256 constant voteSwitchTag = 6;
+    bytes32 constant voteProgram = hex"0761481d357474bb7c4d7624ebd3bdb3d8355e73d11043fc0da3538000000000";
+
+    function countVotes(uint64 slot, bytes memory message) public {
+        SolanaMessage memory parsedMessage = parseSolanaMessage(message);
+        for(uint i = 0; i < parsedMessage.instructions.length; i++) {
+            SolanaInstruction memory instruction = parsedMessage.instructions[i];
+            if (parsedMessage.addresses[instruction.programId] != voteProgram) {
+                continue;
+            }
+
+            // https://docs.rs/solana-vote-program/1.4.13/solana_vote_program/vote_instruction/enum.VoteInstruction.html#variant.Vote
+            bytes memory data = instruction.data;
+            uint tag;
+            uint cursor;
+            (tag, cursor) = parseUint32LE(data, 0);
+            if (tag != voteTag && tag != voteSwitchTag) {
+                continue;
+            }
+
+            uint64 length; uint64 votedSlot;
+            (length, cursor) = parseUint64LE(data, cursor);
+            for(uint j = 0; j < length; j++) {
+                (votedSlot, cursor) = parseUint64LE(data, cursor);
+                if(votedSlot + 32 < slot || slot <= votedSlot)
+                    revert("Voting for invalid slot");
+                slots[slotOffset(votedSlot)].voteCounts++;
+            }
+        }
+    }
+
+    function verifyVote(bytes memory signatures, bytes memory message, uint64 instructionIndex) public pure returns (bool) {
+        uint signaturesCount; uint signaturesOffset;
+        (signaturesCount, signaturesOffset) = parseCompactWord16(signatures, 0);
+
+        SolanaMessage memory parsedMessage = parseSolanaMessage(message);
+        SolanaInstruction memory instruction = parsedMessage.instructions[instructionIndex];
+
+        // To prevent the relayer from hiding votes to avoid challenges, we always check one signature
+        // This might not be needed once stake supermajority checks are possible
+        bytes memory signature; uint cursor;
+        (signature, cursor) = parseSignature(signatures, signaturesOffset);
+        if(! ed25519_valid(signature, message, parsedMessage.addresses[0])) {
```

Comment:
what's this check for?

Reply:
As of now, the challenge only applies to votes, so if a misbehaving relayer removes vote instructions, there would be no vote, and thus no vote authority to verify the signature for - that would allow tampering with messages as long as the votes were removed.
That's what the above comment is meant to explain - should it be updated somehow?